core/network/src/lib.rs

```diff
@@ -17,11 +17,148 @@
+//!
+//! - [Secio](https://github.com/libp2p/specs/tree/master/secio). A TLS-1.2-like protocol but
+//! without certificates. Support for secio will likely be deprecated in the far future.
+//! - [Noise](https://noiseprotocol.org/). Support for noise is very experimental. The details are
+//! very blurry and may change at any moment.
+//!
+//! The following multiplexing protocols are supported:
+//!
+//! - [Mplex](https://github.com/libp2p/specs/tree/master/mplex). Support for mplex will likely
+//! be deprecated in the future.
+//! - [Yamux](https://github.com/hashicorp/yamux).
+//!
+//! ## Substreams
+//!
+//! Once a connection has been established and uses multiplexing, substreams can be opened. When
+//! a substream is open, the **multistream-select** protocol is used to negotiate which protocol to
+//! use on that given substream. In practice, Substrate opens the following substreams:
+//!
+//! - We periodically open an ephemeral substream in order to ping the remote and check whether the
+//! connection is still alive. Failure for the remote to reply leads to a disconnection. This uses
+//! the libp2p ping protocol.
+//! - We periodically open an ephemeral substream in order to ask information from the remote. This
+//! is called [the `identify` protocol](https://github.com/libp2p/specs/tree/master/identify).
+//! - We periodically open ephemeral substreams for Kademlia random walk queries. Each Kademlia
+//! RPC query is done in a new separate substream. This uses the
+//! [standard libp2p Kademlia protocol](https://github.com/libp2p/specs/pull/108).
+//! - We optionally keep a substream alive for all Substrate-based communications. The name of the
+//! protocol negotiated is based on the *protocol ID* passed as part of the network configuration.
+//! This protocol ID should be unique for each chain and prevents nodes from different chains from
+//! connecting to each other. More information below.
+//!
+//! ## The Substrate substream
+//!
+//! Substrate uses a component named the **peerset manager (PSM)**. Through the discovery
+//! mechanism, the PSM is aware of the nodes that are part of the network and decides which nodes
+//! we should perform Substrate-based communications with. For these nodes, we open a connection
+//! if necessary and open a unique substream for Substrate-based communications. If the PSM decides
+//! that we should disconnect a node, then that substream is closed.
+//!
+//! For more information about the PSM, see the *substrate-peerset* crate.
+//!
+//! Note that at the moment there is no mechanism in place to solve the issues that arise where the
+//! two sides of a connection open the unique substream simultaneously. In order to not run into
+//! issues, only the dialer of a connection is allowed to open the unique substream. When the
+//! substream is closed, the entire connection is closed as well. This is a bug, and should be
+//! fixed by improving the protocol.
```

Comment:
You haven't explained why there is this single substrate substream that's long-lived, when everything else is ephemeral substreams. Even if there is a single substream for all of substrate, why not have it be ephemeral too?

Reply:
The answer is technical debt. There's no actual reason, it just hasn't been because the code is a mess to rework.